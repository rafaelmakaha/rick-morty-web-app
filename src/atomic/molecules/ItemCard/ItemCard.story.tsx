import React from 'react'
import { Meta, Story } from '@storybook/react'
import ItemCard, { IItemCard } from '.'


export default {
  title: 'Molecules/ItemCard'
} as Meta

const Template: Story<IItemCard> = (args) => <ItemCard {...args} />

export const Normal = Template.bind({})

const argTypes = {
  color: {
    control: 'color'
  },
  char: {
    control: 'select',
    options: ['Rick', 'Morty', 'Jerry'],
    mapping: {
      Rick: {
        "name": "Rick Sanchez",
        "status": "Alive",
        "species": "Human",
        "gender": "Male",
        "image": "https://rickandmortyapi.com/api/character/avatar/1.jpeg"
      },
      Morty: {
        "name": "Morty Smith",
        "status": "Alive",
        "species": "Human",
        "gender": "Male",
        "image": "https://rickandmortyapi.com/api/character/avatar/2.jpeg"
      },
      Jerry: {
        "name": "Jerry Smith",
        "status": "Alive",
        "species": "Human",
        "gender": "Male",
        "image": "https://rickandmortyapi.com/api/character/avatar/5.jpeg"
      }
    },
  }
}

const defaultArgs: Omit<IItemCard> = {
  color: "#506506",
  char: {
    "name": "Rick Sanchez",
    "status": "Alive",
    "species": "Human",
    "image": "https://rickandmortyapi.com/api/character/avatar/1.jpeg"
  }
}

Normal.args = { ...defaultArgs }
Normal.argTypes = argTypes

