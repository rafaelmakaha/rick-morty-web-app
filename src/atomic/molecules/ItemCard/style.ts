import styled from 'styled-components'

export const Container = styled.div`
    border: 2px;
    border-radius: 10px;
    background-color: ${(props) => props.color || "#506506"};
    display: flex;
    flex-direction: column;
    height: auto;
    padding: 4px;
    width: auto;
`

export const Image = styled.img`
    border: 2px;
    border-radius: 5px;
    object-fit: fill;
`