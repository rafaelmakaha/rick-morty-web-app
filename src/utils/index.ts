export * from './assignTestId'
export * from './withProviders'
