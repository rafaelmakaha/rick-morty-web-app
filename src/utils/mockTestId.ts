type componentType = 'div' | 'text' | 'input' | 'button' | 'img'

export const assignTestId = (
  componentType: componentType,
  componentId: string
) =>
  process.env.NODE_ENV !== 'production'
    ? {
        'data-testid': `${componentType}_${componentId}`
      }
    : {}

export const mockTestID = (
  componentType: componentType,
  componentId: string
): string => assignTestId(componentType, componentId)['data-testid'] ?? ''
