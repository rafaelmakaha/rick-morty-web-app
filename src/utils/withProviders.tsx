import React from 'react'

import { ThemeProvider } from 'styled-components'

import { theme } from '../tokens'

export const StorybookProviders: React.FC = ({ children }) => (
  <ThemeProvider theme={theme}>{children}</ThemeProvider>
)

export const providerDecorator = (Story: any) => (
  <StorybookProviders>
    <Story />
  </StorybookProviders>
)
