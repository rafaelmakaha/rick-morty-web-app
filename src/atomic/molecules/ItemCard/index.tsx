import React from "react";
import { ICharacter } from "../../../services/types";
import CardTitle from "../../atoms/CardTitle";
import ItemDescription from "../../atoms/ItemDescription";
import { Container, Image } from "./style";

export interface IItemCard {
    testID?: string
    char: ICharacter
}

const ItemCard: React.FC<IItemCard> = ({ testID = 'ItemCard', char}) => (
    <Container>
        <CardTitle>
            {char?.name}
        </CardTitle>
        <ItemDescription>
            {char?.status}
        </ItemDescription>
        <ItemDescription>
            {char?.species}
        </ItemDescription>
        <ItemDescription>
            {char?.gender}
        </ItemDescription>
        <Image alt={char?.name} src={char?.image}/>
    </Container>
)

export default ItemCard