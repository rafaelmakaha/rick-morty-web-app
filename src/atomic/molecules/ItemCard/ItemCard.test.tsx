import React from 'react'
import { render, cleanup } from '@testing-library/react'
import { mockTestID } from '../../../utils'
import ItemCard, { IItemCard } from '.'

describe('Atoms/ItemCard', () => {
    const props: IItemCard = {
        testID: 'ItemCard',
        char: {
            id: 0,
            name: 'Character 0',
            status: 'Some Status',
            species: 'Some Species',
            type: 'Some type',
            gender: 'Some gender',
            image: 'Some url'
        }
    }

    const renderComponent = () => render(<ItemCard {...props} />)

    let component = renderComponent()

    beforeEach(() => {
        jest.clearAllMocks()
        cleanup()
        component = renderComponent()
    })

    const el_container = mockTestID('div', props.testID!)


    it(`should render component #${el_container}`, () => {
        const sut = component.findByTestId(el_container)
        expect(sut).toBeTruthy()
    })

    it('should render snapshot', () => {
        expect(component.container.firstChild).toMatchSnapshot()
    })
})
