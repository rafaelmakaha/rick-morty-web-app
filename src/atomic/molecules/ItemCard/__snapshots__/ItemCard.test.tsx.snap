// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Atoms/ItemCard should render snapshot 1`] = `
<div
  class="style__Container-sc-1ll6tf3-0 ebsKbC"
>
  <h2
    class="style__Title-sc-1wrd5mx-0 fqesxN"
  >
    Character 0
  </h2>
  <p
    class="style__Text-sc-bm4pk8-0 kktcfB"
  >
    Some Status
  </p>
  <p
    class="style__Text-sc-bm4pk8-0 kktcfB"
  >
    Some Species
  </p>
  <p
    class="style__Text-sc-bm4pk8-0 kktcfB"
  >
    Some gender
  </p>
  <img
    alt="Character 0"
    class="style__Image-sc-1ll6tf3-1 cxnJHb"
    src="Some url"
  />
</div>
`;
